interop-reactiveStreams/shared/src/main/scala/scalaz/zio/interop/reactiveStreams/StreamPublisher.scala

```diff
@@ -0,0 +1,62 @@
+package scalaz.zio.interop.reactiveStreams
+
+import org.reactivestreams.{ Publisher, Subscriber, Subscription }
+import scalaz.zio.{ Queue, Runtime, Task, UIO, ZIO }
+import scalaz.zio.stream.{ Sink, Stream, Take }
+
+class StreamPublisher[E <: Throwable, A](
+  stream: Stream[Any, E, A],
+  runtime: Runtime[Any]
+) extends Publisher[A] {
+
+  override def subscribe(subscriber: Subscriber[_ >: A]): Unit =
+    if (subscriber == null) {
+      throw new NullPointerException("Subscriber must not be null.")
+    } else {
+      runtime.unsafeRunAsync_(
+        for {
+          demand  <- Queue.unbounded[Long]
+          _       <- Task(subscriber.onSubscribe(new QSubscription(subscriber, demand)))
+          control = Stream.fromQueue(demand).flatMap(n => Stream.unfold(n)(n => if (n > 0) Some(((), n - 1)) else None))
+          fiber   <- stream.toQueue().use(takesToCallbacks(subscriber, _, control, demand)).fork
+          // reactive streams rule 3.13
+          _ <- (demand.awaitShutdown *> fiber.interrupt).fork
+        } yield ()
+      )
+    }
+
+  private def takesToCallbacks(
+    subscriber: Subscriber[_ >: A],
+    sourceQ: Queue[Take[E, A]],
+    control: Stream[Any, Nothing, Unit],
+    demandQ: Queue[Long]
+  ): Task[Unit] =
+    Stream
+      .fromQueue(sourceQ)
+      // handle initially failed source before receiving demand
+      .peel(Sink.readWhile[Take[E, A]](_.isInstanceOf[Take.Fail[E]]))
```

Comment:
```scala
readWhile {
  case Take.Fail(_) => true
  case _            => false
}
```

Reply:
Can do. As a matter of style, or are there additional advantages? TBO while I'd prefer pattern matching in general, I think in this case _.isInstanceOf is more compact.

Reply:
Well, `isInstanceOf` is not referentially transparent; this project is liberal in use of unsafe constructs for performance, but only when absolutely necessary. In this case, I believe the generated bytecode should be the same.

In terms of compactness, it could be useful to add a `def isFailure: Boolean` to `Take` itself and use it here.

Reply:
done